src/sinks/opendal_common.rs

```diff
@@ -0,0 +1,281 @@
+//! opendal_common provide real sink supports for all opendal based services.
+//!
+//! # TODO
+//!
+//! opendal service now only support very basic sink features. To make it
+//! useful, we need to add the following features:
+//!
+//! - Error handling
+//! - Limitation
+
+use std::{fmt, num::NonZeroUsize, task::Poll};
+
+use bytes::Bytes;
+use codecs::encoding::Framer;
+use futures::{stream::BoxStream, StreamExt};
+use opendal::Operator;
+use snafu::Snafu;
+use tower::Service;
+use tracing::Instrument;
+use vector_common::{
+    finalization::{EventStatus, Finalizable},
+    request_metadata::{MetaDescriptive, RequestMetadata},
+};
+use vector_core::{
+    internal_event::CountByteSize,
+    sink::StreamSink,
+    stream::{BatcherSettings, DriverResponse},
+    ByteSizeOf,
+};
+
+use crate::{
+    codecs::{Encoder, Transformer},
+    event::{Event, EventFinalizers},
+    internal_events::SinkRequestBuildError,
+    sinks::{
+        util::{
+            metadata::RequestMetadataBuilder, partitioner::KeyPartitioner,
+            request_builder::EncodeResult, Compression, RequestBuilder, SinkBuilderExt,
+        },
+        BoxFuture,
+    },
+};
+
+pub struct OpenDalSink<Svc> {
```

Comment:
Can we add a doc comment to all our public objects... I realise that there are huge areas of Vector that are currently lacking in this, but we are working on it.

Reply:
DONE.